juju/juju
# The firewaller creates ingress rules for remote applications

## Description of change

The firewaller is updated to allow ingress from remote models so that remote relations can function.
Core components:
- new RemoteRelations facade method IngressSubnetsForRelation
- fix to state lifecycle watcher to properly filter members on initialisation
- remove the automatic expose to the world behaviour from add relation api
- update firewaller to listen to remote relations

The watcher fix to to plug a hole where the initial watcher elements are not properly filtered because only the member query was being used, not the filter method.

When a relation is added, the add relation API no longer needs to open the applications to the world as the firewaller now takes care of ingress management.

The firewaller is not complete - it reacts to new remote relations but the necessary network updates in the model may not have been completed. So the firewaller will fall back to using 0.0.0.0/0. A follow up will add a watcher to the firewaller to handle things better.

The IngressSubnetsForRelation API for now just returns all CIDRs from the source model. The next step will be to only return CIDRs from machines hosting units of the remote application. And then next step will be to uses the spaces model to return only CIDRs to which the remote endpoints are bound.

The firewaller fallback to 0.0.0.0/0 is also needed because only AWS and MAAS currently implement the environ NetworkInterface. Azure, GCE, Openstack need updating before the fallback can be dropped.

## QA steps

bootstrap aws
deploy mediawiki and mysql in separate models
relate mediawiki and mysql
check the mysql security group to see that it allows ingress from 0.0.0.0/0
